streams/src/main/java/org/apache/kafka/streams/query/MultiVersionedKeyQuery.java

```diff
@@ -0,0 +1,127 @@
+    private final Optional<Instant> toTime;
+    private final boolean isAscending;
+
+    private MultiVersionedKeyQuery(final K key, final Optional<Instant> fromTime, final Optional<Instant> toTime, final boolean isAscending) {
+        this.key = Objects.requireNonNull(key);
+        this.fromTime = fromTime;
+        this.toTime = toTime;
+        this.isAscending = isAscending;
+    }
+
+      /**
+       * Creates a query that will retrieve the set of records identified by {@code key} if any exists
+       * (or {@code null} otherwise).
+       * @param key The key to retrieve
+       * @throws NullPointerException if @param key is null
+       * @param <K> The type of the key
+       * @param <V> The type of the value that will be retrieved
+       * @throws NullPointerException if @param key is null
+       */
+    public static <K, V> MultiVersionedKeyQuery<K, V> withKey(final K key) {
+        return new MultiVersionedKeyQuery<>(key, Optional.empty(), Optional.empty(), true);
+    }
+
+    /**
+     * Specifies the starting time point for the key query.
+     * <pre>
+     * The key query returns all the records that are still existing in the time range starting from the timestamp {@code fromTime}. There can be records which have been inserted before the {@code fromTime}
+     * and are valid in the query specified time range (the whole time range or even partially). The key query in fact returns all the records that have NOT become tombstone at or after {@code fromTime}.
+     * </pre>
+     * @param fromTime The starting time point
+     * If @param fromTime is null, will be considered as negative infinity
+     */
+    public MultiVersionedKeyQuery<K, V> fromTime(final Instant fromTime) {
+        if (fromTime == null) {
+            return new MultiVersionedKeyQuery<>(key, Optional.empty(), toTime, true);
+        }
+        return new MultiVersionedKeyQuery<>(key, Optional.of(fromTime), toTime, true);
+    }
+
+    /**
+     * Specifies the ending time point for the key query.
+     * The key query returns all the records that have timestamp <= {@code toTime}.
+     * @param toTime The ending time point
+     * If @param toTime is null, will be considered as positive infinity
+     */
+    public MultiVersionedKeyQuery<K, V> toTime(final Instant toTime) {
+        if (toTime == null) {
+            return new MultiVersionedKeyQuery<>(key, fromTime, Optional.empty(), true);
+        }
+        return new MultiVersionedKeyQuery<>(key, fromTime, Optional.of(toTime), true);
+    }
+
+    /**
+     * Specifies the order of the returned records by the query as descending by timestamp.
+     */
+    public MultiVersionedKeyQuery<K, V> withDescendingTimestamps() {
+        return new MultiVersionedKeyQuery<>(key, fromTime, toTime, false);
+    }
+
+    /**
+     * Specifies the order of the returned records by the query as ascending by timestamp.
+     */
+    public MultiVersionedKeyQuery<K, V> withAscendingTimestamps() {
```

Comment:
> Just thinking about consistency again -- for `RangeQuery` (Hanyu's KIP) we did not add this, because it's the default anyway. Should we remove it (and update the KIP accordingly) -- or update Hanyu's KIP to add something similar?

Mmmm. I added that based the KIP discussions and feedback in order to have more code readability.

Reply:
Yeah. I am not against to add it. We can also consider it a "slip" in @hanyuzheng7's KIP and thus could also add it there.